### PR TITLE
[SPARK-46723][CONNECT][SCALA] Make addArtifact retryable

### PR DESCRIPTION
**Pull request by cdkrot**
### What changes were proposed in this pull request?

Make addArtifact API retrying on errors.

Note this is safe operation since addArtifact is idempotent operation (https://github.com/apache/spark/pull/43314)

### Why are the changes needed?

For the same reasons as we make other API retryable.

### Does this PR introduce _any_ user-facing change?

Yes

### How was this patch tested?

Added test.

Testing by hand against custom spark server.

### Was this patch authored or co-authored using generative AI tooling?

No & never

### CODE REVIEW

**Comment by cdkrot** [comment]
cc @vicennial @nija-at @hvanhovell

**Review by vicennial — APPROVED** [review]
LGTM!

File: connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala [hunk 1]
```diff
@@ -451,6 +451,33 @@ class SparkConnectClientSuite extends ConnectFunSuite with BeforeAndAfterEach {
...
+  test("ArtifactManager retries errors") {
+    var attempt = 0
+
+    startDummyServer(0)
+    client = SparkConnectClient
+      .builder()
+      .connectionString(s"sc://localhost:${server.getPort}")
+      .interceptor(new ClientInterceptor {
+        override def interceptCall[ReqT, RespT](
+                                                 methodDescriptor: MethodDescriptor[ReqT, RespT],
```
Comment (hvanhovell) [inline connector/connect/client/jvm/src/test/scala/org/apache/spark/sql/connect/client/SparkConnectClientSuite.scala:463]: NIT style

**Review by hvanhovell — APPROVED** [review]
LGTM

**Comment by cdkrot** [comment]
Tests green!

**Review by HyukjinKwon — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**